crates/extensions/c8y_mapper_ext/src/config_operations.rs

```diff
@@ -0,0 +1,1338 @@
+use tedge_api::messages::ConfigUpdateCmdPayload;
+use tedge_api::mqtt_topics::Channel;
+use tedge_api::mqtt_topics::ChannelFilter::Command;
+use tedge_api::mqtt_topics::ChannelFilter::CommandMetadata;
+use tedge_api::mqtt_topics::EntityFilter::AnyEntity;
+use tedge_api::mqtt_topics::EntityTopicId;
+use tedge_api::mqtt_topics::MqttSchema;
+use tedge_api::mqtt_topics::OperationType;
+use tedge_api::Jsonify;
+use tedge_downloader_ext::DownloadRequest;
+use tedge_downloader_ext::DownloadResult;
+use tedge_mqtt_ext::Message;
+use tedge_mqtt_ext::MqttMessage;
+use tedge_mqtt_ext::QoS;
+use tedge_mqtt_ext::TopicFilter;
+use tedge_utils::file::create_directory_with_defaults;
+use tedge_utils::file::create_file_with_defaults;
+use tracing::log::info;
+use tracing::log::warn;
+
+pub fn config_snapshot_topic_filter(mqtt_schema: &MqttSchema) -> TopicFilter {
+    [
+        mqtt_schema.topics(AnyEntity, Command(OperationType::ConfigSnapshot)),
+        mqtt_schema.topics(AnyEntity, CommandMetadata(OperationType::ConfigSnapshot)),
+    ]
+    .into_iter()
+    .collect()
+}
+
+pub fn config_update_topic_filter(mqtt_schema: &MqttSchema) -> TopicFilter {
+    [
+        mqtt_schema.topics(AnyEntity, Command(OperationType::ConfigUpdate)),
+        mqtt_schema.topics(AnyEntity, CommandMetadata(OperationType::ConfigUpdate)),
+    ]
+    .into_iter()
+    .collect()
+}
+
+impl CumulocityConverter {
+    /// Convert c8y_UploadConfigFile SmartREST request to ThinEdge config_snapshot command.
+    /// Command ID is generated here, but it should be replaced by c8y's operation ID in the future.
+    pub fn convert_config_snapshot_request(
+        &self,
+        smartrest: &str,
+    ) -> Result<Vec<Message>, CumulocityMapperError> {
+        if !self.config.capabilities.config_snapshot {
+            warn!("Received a c8y_UploadConfigFile operation, however, config_snapshot feature is disabled");
+            return Ok(vec![]);
+        }
+
+        let snapshot_request = SmartRestConfigUploadRequest::from_smartrest(smartrest)?;
+        let target = self
+            .entity_store
+            .get_by_external_id(&snapshot_request.device.clone().into())
+            .ok_or_else(|| UnknownDevice {
+                device_id: snapshot_request.device.to_string(),
+            })?;
+
+        let cmd_id = nanoid!();
+        let channel = Channel::Command {
+            operation: OperationType::ConfigSnapshot,
+            cmd_id: cmd_id.clone(),
+        };
+        let topic = self.mqtt_schema.topic_for(&target.topic_id, &channel);
+        let external_id: String = target.external_id.clone().into();
+
+        // Replace '/' with ':' to avoid creating unexpected directories in file transfer repo
+        let tedge_url = format!(
+            "http://{}/tedge/file-transfer/{}/config_snapshot/{}-{}",
+            &self.config.tedge_http_host,
+            external_id,
+            snapshot_request.config_type.replace('/', ":"),
```

Comment:
Our config_type allows to use `/` (it's even a kind of default when user doesn't give specific type, then using file path as type). However, since `/` is considered directory separator, so somehow need to escape. Then I introduce replacing with `:` .

If we don't escape, when plugin uploads to `http://localhost:8888/tedge/file-transfer/test-device/config_snapshot//etc/tedge/tedge.toml-1234` for the type `/etc/tedge/tedge.toml`, the file will be stored `/var/tedge/test-device/config_snapshot/etc/tedge/tedge.toml-1234`. Too many directories.